server/src/com/mirth/connect/server/userutil/DestinationSet.java

```diff
@@ -7,162 +8,221 @@
-     * 
-     * @param metaDataIdOrConnectorNames
-     *            A collection of integers representing the metaDataId of a destination connectors,
-     *            or the actual destination connector names. JavaScript arrays can be used.
-     * @return A boolean indicating whether at least one destination connector was actually removed
-     *         from processing for this message.
-     */
-    public boolean removeAllExcept(Collection<Object> metaDataIdOrConnectorNames) {
-        if (metaDataIds != null) {
-            Set<Integer> set = new HashSet<Integer>();
-
-            for (Object metaDataIdOrConnectorName : metaDataIdOrConnectorNames) {
-                Integer metaDataId = convertToMetaDataId(metaDataIdOrConnectorName);
-
-                if (metaDataId != null) {
-                    set.add(metaDataId);
-                }
-            }
-
-            return metaDataIds.retainAll(set);
-        }
-
-        return false;
-    }
-
-    /**
-     * Stop all destinations from being processed for this message. This does NOT mark the source
-     * message as FILTERED.
-     * 
-     * @return A boolean indicating whether at least one destination connector was actually removed
-     *         from processing for this message.
-     */
-    public boolean removeAll() {
-        if (metaDataIds != null && metaDataIds.size() > 0) {
-            metaDataIds.clear();
-            return true;
-        }
-
-        return false;
-    }
-
-    private Integer convertToMetaDataId(Object metaDataIdOrConnectorName) {
-        if (metaDataIdOrConnectorName != null) {
-            if (metaDataIdOrConnectorName instanceof Number) {
-                return ((Number) metaDataIdOrConnectorName).intValue();
-            } else if (metaDataIdOrConnectorName.getClass().getName().equals("org.mozilla.javascript.NativeNumber")) {
-                return (Integer) Context.jsToJava(metaDataIdOrConnectorName, int.class);
-            } else if (destinationIdMap != null) {
-                return destinationIdMap.get(metaDataIdOrConnectorName.toString());
-            }
-        }
-
-        return null;
-    }
-}
\ No newline at end of file
+ package com.mirth.connect.server.userutil;
+
+ import java.util.Collection;
+ import java.util.Collections;
+ import java.util.HashSet;
+ import java.util.Iterator;
+ import java.util.Map;
+ import java.util.Optional;
+ import java.util.Set;
+ import java.util.stream.Collectors;
+ 
+ import org.mozilla.javascript.Context;
+ 
+ import com.mirth.connect.donkey.server.Constants;
+ import com.mirth.connect.userutil.ImmutableConnectorMessage;
+ 
+ /**
+  * Utility class used in the preprocessor or source filter/transformer to prevent the message from
+  * being sent to specific destinations.
+  */
+ public class DestinationSet implements Set<Integer> {
+ 
+     private Map<String, Integer> destinationIdMap = Collections.emptyMap();
+     private Set<Integer> metaDataIds;
```

Comment:
I assume you are initializing `destinationIdMap` here in case it fails to set in the constructor. Can we initialize `metaDataIds` here as well instead of adding it to the catch block of the constructor? I think further discussion is needed on whether or not it's a good idea to change the behavior of DestinationSet not allowing additions. I think for the time being it can be initialized with `Collections.emptySet()`.